docs/design/1161-standardize-stdlib.md

```diff
@@ -0,0 +1,237 @@
+* `join`
+* `json_decode`
+* `json_path`
+* `nonsensitive`
+* `replace`
+* `split`
+* `to_lower`
+* `to_upper`
+* `trim`
+* `trim_prefix`
+* `trim_space`
+* `trim_suffix`
+* `yaml_decode`
+
+These identifiers were added organically over time; there has been little
+thought to consistency across all identifiers.
+
+While 19 identifiers is manageable and possible for a user to remember the most
+important ones, if we don't establish a way to standardise existing and new
+identifiers, the inconsistency between identifiers will only grow over time as
+new identifiers are added and become harder to understand.
+
+This proposal aims to:
+
+* Establish a convention for how to name identifiers in the standard library.
+* Propose a way to align existing identifiers with the new convention without
+  breaking backwards compatibility.
+
+For the sake of demonstrating the convention, new identifiers are provided as
+examples, but are out of scope of this proposal and should not be considered
+accepted if this proposal is accepted.
+
+## Proposal
+
+The standard library will introduce namespaces for identifiers that group
+similar identifiers together:
+
+* Array functions
+    * `array.concat` (previously `concat`)
+* Conversion functions
+    * `convert.nonsensitive` (previously `nonsensitive`)
+* Decoding functions
+    * `decode.base64` (previously `decode_base64`)
+    * `decode.json` (previously `decode_json`)
+    * `decode.yaml` (previously `decode_yaml`)
+* OS-related functions
+    * `os.env` (previously `env`)
+* String functions
+    * `string.format` (previously `format`)
+    * `string.join` (previously `join`)
+    * `string.replace` (previously `replace`)
+    * `string.split` (previously `split`)
+    * `string.to_lower` (previously `to_lower`)
+    * `string.to_upper` (previously `to_upper`)
+    * `string.trim` (previously `trim`)
+    * `string.trim_prefix` (previously `trim_prefix`)
+    * `string.trim_space` (previously `trim_space`)
+    * `string.trim_suffix` (previously`trim_suffix`)
+
+For identifiers where a namespace has been introduced, the old identifier will
+be marked deprecated for removal for the next major release.
+
+Some identifiers have not been given a namespace because I could not easily
+identify one: `constants`, `coalesce`, and `json_path`. Until a namespace is
+identified (if ever), these identifiers will not be deprecated.
+
+New identifiers **should** be given a namespace; namespacing makes it easier
+for the same identifier to exist in two different namespaces, such as
+`string.join` (for joining multiple strings) alongside a hypothetical
+`path.join` (for joining file system paths).
+
+Introducing a new namespace should be done as a proposal; ideally proposing
+multiple identifiers that would comprise that namespace. Introducing new
+identifiers in an existing namespace should also be done as a proposal, much
+like today.
+
+I've elected for namespaces to use dot separation (`NAMESPACE.IDENTIFIER`) as
```

Comment:
Worth mentioning in the proposal: The new namespaces may collide with existing custom components but it should not be a breaking change because the custom components shadow the stdlib and because of the aliasing.